File: 2025-04-03/Diameter_Of_Binary_Tree.java
package amazon;

public class Diameter_Of_Binary_Tree {
    public static int ans;
    public static int solve(TreeNode root) {
        if(root == null) return 0;

        int left = diameterOfBinaryTree(root.left);
        int right = diameterOfBinaryTree(root.right);

        ans = Math.max(ans, left + right);

        return 1 + Math.max(left, right);
    }
    public static int diameterOfBinaryTree(TreeNode root) {
        ans = 0;
        solve(root);
        return ans;
    }
    public static void main(String[] args) {
        Integer[] arr = new Integer[]{3,1,null,null,2};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        int ans = diameterOfBinaryTree(root);
        System.out.println(ans);
    }
}

File: 2025-10-01/Diameter_Of_Binary_Tree.py
from collections import deque
from typing import Optional, List

from TreeNode import TreeNode
from Tree import Tree


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return True

        ans = [0]
        def dfs(root):
            if not root:
                return 0

            l = dfs(root.left)
            r = dfs(root.right)

            ans[0] = max(ans[0], l + r)
            return 1 + max(l, r)

        dfs(root)
        return ans[0]


sol = Solution()
tree = Tree()
node = tree.array_to_tree([3,1,None,None,2])
print(sol.diameterOfBinaryTree(node))
